tests/specs/mcd/verification.k

```diff
@@ -138,6 +160,23 @@ module LEMMAS-MCD-HASKELL [symbolic, kore]
     rule #range(M:Memory [ K <- V ] [ K' := VS':ByteArray ], START, WIDTH) => #range(M [ K := (V : VS')        ], START, WIDTH)  requires K' +Int #sizeByteArray(VS') ==Int START +Int WIDTH andBool START <=Int K andBool 0 <Int WIDTH andBool 0 <=Int START
                                                                                                                                   andBool K +Int 1                    ==Int K'                                                                                [concrete(K, K', V, VS'), simplification]
 
+    rule { (M:Map [K <- V1]) #Equals (M [K <- V2]) } => { V1 #Equals V2 }
+         [simplification]
+
+    rule { M:Map [K <- V1] #Equals M } =>
+         { true #Equals K in_keys(M) andBool V1 ==K #lookup (M, K) }
+         [simplification]
```

Comment:
The condition `V1 ==K #lookup(M, K)` either needs to be generalized, or it needs to be moved to a side-condition.

This equality is actually an implication, the LHS of this rule implies that the RHS is true, but the RHS does not imply that the LHS is true. A counterexample would be a map `M` where the key `K` is bound to something that is not an integer.

Reply:
In general, to write a rule like `rule LHS => RHS requires REQ [simplification]`, you _must_ take care that `REQ implies (LHS == RHS)` is true.

Here, `REQ` is empty (so is just `true`), and you do not have `LHS == RHS`, you have `LHS implies RHS`.

Reply:
This is interesting, and correct. But isn't this an equivalence in the context of KEVM, given that all maps are assumed to have the codomain `Int` (this I infer from the non-int simplification rule of `#lookup` having the annotation "Impossible")? How about `V1 ==K M[K]`?

Reply:
I will need to re-run all tests after changing this because I cannot remember for which test this was relevent.

Reply:
KEVM uses the `Map` type for other things than Storage though, where we don't necessarily have the same restrictions.

So either the lemma should be lowered to operate over `Map` type directly, or made an implication instead.

Reply:
What would it mean to make it an implication?

Reply:
I realized I'm not quite understanding the other variant either - lowered to operate over `Map`?

Reply:
The simplification is not needed in the end, all of the relevant tests pass without it.